### PR TITLE
Support Long Timers ⏳📈🚀🚀🚀

### LINKED ISSUE #325: Support Long Timers
This PR is to support timers longer than 7 days using the [createTimer[(https://docs.microsoft.com/en-us/javascript/api/durable-functions/durableorchestrationcontext?view=azure-node-latest#durable-functions-durableorchestrationcontext-createtimer) API.

To do this, we'll need changes both at the extension-level, and at the DF SDK level.
At the extension level:
* We include in the SDK payload a new field indicating the current maximum time for an individual timer
* Given that we're modifying the SDK payload, we should increase the SDK protocol version number
In the SDK:
* We should increase the SDK protocol version number to support the new payload
* We need to modify our TimerTask state machine to fetch the right history event, the one corresponding to the "final sub-timer"

### PR DESCRIPTION
**Pull request by hossam-nasr**
Resolves #325. This PR leverages these changes in the extension: https://github.com/Azure/azure-functions-durable-extension/pull/2134 to allow JS orchestrators to have timers longer than 6 days. Internally, when user code calls `CreateTimer`, if it is more than 6 days, a `LongTimerTask`, which is essentially a `WhenAll` task of several small sub-timers, is created instead of an atomic `DFTimerTask`. 

Now, the JS SDK will become the first OOProc SDK to support long timers 🥳 This also unblocks #284

### CODE REVIEW

File: src/durableorchestrationcontext.ts [hunk 1]
```diff
@@ -293,9 +334,25 @@ export class DurableOrchestrationContext {
...
+        ) {
+            return new LongTimerTask(
+                false,
+                timerAction,
+                this,
+                this.taskOrchestratorExecutor,
+                this.maximumShortTimerDuration,
+                this.longRunningTimerIntervalDuration
+            );
+        }
```
Comment (davidmrdavid) [inline src/durableorchestrationcontext.ts:353]: I think it'd be wise to protect against the case where ReplaySchema is V3 but the maximumShortTimerDuration and/or the longRunningTimerIntervalDuration are undefined.

Can we add a check here that throws an exception if that's the case? Otherwise, we risk things going wrong in the extension-level.

File: src/task.ts [hunk 2]
```diff
@@ -377,6 +379,107 @@ export class WhenAnyTask extends CompoundTask {
...
+        id: TaskID,
+        action: CreateTimerAction,
+        orchestrationContext: DurableOrchestrationContext,
+        executor: TaskOrchestrationExecutor,
+        maximumTimerLength: moment.Duration,
+        longRunningTimerIntervalDuration: moment.Duration
+    ) {
+        const currentTime = moment(orchestrationContext.currentUtcDateTime);
+        const finalFireTime = moment(action.fireAt);
+        const durationTillFire = moment.duration(finalFireTime.diff(currentTime));
```
Comment (davidmrdavid) [inline src/task.ts:411]: nit: just for style's sake, can we change `durationTillFire` to `durationUntilFire`? Just trying to avoid the abbreviations :)

Alternatively, I also like `timerLength` for this variable.

**Review by davidmrdavid — COMMENTED** [review]
Good job, I think this is close to merge. Left some comments. My main blocker at this point is that I would prefer not having the timer metadata be exposed in the orchestration context class :-)

File: src/durableorchestrationcontext.ts (hunk 1 above)
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts:353]: I was actually going to ask for feedback regarding this `if` statement. Should we even be checking for schema version? If the schema is less than `V3`, an exception would be thrown in the extension, regardless of whether we create a `LongTimerTask` or a `DFTimerTask`. 

Also, I was going to ask what should we do in the case that `maximumShortTimerDuration` is defined but not `longRunningTimerIntervalDuration`? Should we also complain/throw an exception? Just ignore and create a `DFTimerTask`? Default to using `maximumShortTimerDuration` as the `longRunningTimerIntervalDuration`?

File: src/task.ts (hunk 2 above)
Comment (hossam-nasr) [reply src/task.ts]: Done, removed the abbreviation :)

File: src/durableorchestrationcontext.ts [hunk 3]
```diff
@@ -101,6 +114,34 @@ export class DurableOrchestrationContext {
...
+
+    /**
+     * Gets the current schema version that this execution is
+     * utilizing, based on negotiation with the extension.
+     *
+     * Different schema versions can allow different behavior.
+     * For example, long timers are only supported in schema version >=3
+     */
+    public schemaVersion: ReplaySchema;
+
```
[the remark this answers is not in the document]
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: I made them private and readonly

File: src/durableorchestrationcontext.ts (hunk 1 above)
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: Add the check! On second thought, I actually think we need to wrap this in a check for the replay version if we want to throw this error. We don't want to complain about timer-related metadata not being set if schema version < 3 is being used anyway.

The alternative I guess would be to ignore this case entirely and let it error out in the extension, but I think throwing the error here is more helpful.

Also, would appreciate thoughts on the error message 😛

File: src/durableorchestrationcontext.ts [hunk 4]
```diff
@@ -293,9 +334,28 @@ export class DurableOrchestrationContext {
...
     public createTimer(fireAt: Date): TimerTask {
-        const newAction = new CreateTimerAction(fireAt);
-        const task = new DFTimerTask(false, newAction);
-        return task;
+        const timerAction = new CreateTimerAction(fireAt);
+        const durationUntilFire = moment.duration(moment(fireAt).diff(this.currentUtcDateTime));
+        if (this.schemaVersion >= ReplaySchema.V3) {
+            if (!this.maximumShortTimerDuration || !this.longRunningTimerIntervalDuration) {
+                throw Error(
+                    "Replay schema version >= V3 is being used, but one or more of the properties `maximumShortTimerDuration` and `longRunningTimerIntervalDuration` are not defined. This is likely an issue with the Extension."
```
Comment (davidmrdavid) [inline src/durableorchestrationcontext.ts:342]: It would be good to include actionable feedback for errors like this. In this case, let's ask users to report the bug :)

```suggestion
                    "A framework-internal error was detected: replay schema version >= V3 is being used, but one or more of the properties `maximumShortTimerDuration` and `longRunningTimerIntervalDuration` are not defined. This is likely an issue with the Durable Functions Extension. Please report this bug here: https://github.com/Azure/azure-functions-durable-js/issues "

```

Comment (davidmrdavid) [reply src/durableorchestrationcontext.ts:342]: Additionally, it would be good if you could log the values for `maximumShortTimerDuration` and `longRuninngTimerIntervalDuration`. It's a bit verbose, but it will help us greatly if a bug is filed.

**Review by davidmrdavid — COMMENTED** [review]
We're almost there!

File: src/durableorchestrationcontext.ts (hunk 1 above)
Comment (davidmrdavid) [reply src/durableorchestrationcontext.ts]: Left some comments on the error message in my latest review!

File: src/durableorchestrationcontext.ts (hunk 4 above)
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: I could log them, but they won't be the raw JSON values that we get from the extension. Those are parsed [here](https://github.com/Azure/azure-functions-durable-js/blob/e2f32c0a65821e53bc24c53b704e1ac3811f7edb/src/durableorchestrationcontext.ts#L58) into either a `moment.Duration` object, or `undefined`, so if we hit this line, it pretty much means they were `undefined`.

Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: Although I guess it would help decipher which one was undefined, or if it was both

Comment (davidmrdavid) [reply src/durableorchestrationcontext.ts]: Yup, I suggested this to figure out which one was undefined :)

Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: Added the logging

**Review by davidmrdavid — COMMENTED** [review]
This looks pretty much ready to me. Just 2 more things:
1. I realized last night this was missing a test-case. Can you please add the appropiate long timer test cases using the following segment of timer tests as an example? https://github.com/Azure/azure-functions-durable-js/blob/3182ada64ddcd7acfeef9bf6b8eca34d6deeaf75/test/integration/orchestrator-spec.ts#L1536

For the above, you'll probably have to modify the `TestHistories` class to include a new history for long timers. That will be a little manual, but essentially you need to mock the expected History for a simple long timer scenario. My tip here is not to overcomplicate it, just mock the history for an orchestrator that schedules 1 long timer with, say, 2 sub-timers, and that's it.

As for how many long timer test cases you need, I'd say 2: one that tests that the orchestrator waits for the sub-timers, and another that tests that the orchestrator proceeds when all sub-timers finish.

2. Can you do a manual check testing that long timers can be cancelled successfully? Now that timers can be arbitrarily long, it is important to make sure they can be cancelled at-will. The easiest way to test for cancellable timers it to utilize the [human in the loop](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-overview?tabs=javascript#human) sample and have the activity in it always finish immediately. See the linked sample for more clarity. Just want to make sure this edge-case is at least manually tested as well. Thanks!

**Review by hossam-nasr — COMMENTED** [review]
Added unit tests and also manually tested and verified that long timers get canceled :)

**Review by davidmrdavid — APPROVED** [review]
The time has come ... 
to merge this PR 🧙‍♂️ 

✨ _it looks good to me_ ✨
